readme.md
# Django LDAP

This library provides LDAP3 integration for Django, its purpose is to
wrap the ldap3 library and provide useful helpers for searching and
manipulating a given directory.

This library intends on becoming a full-fledged authentication backend
for Django in coming updates.

## Requirements
* Django 2.2+
* Python 3.7+

## Installation
1. Install the library
~~~sh
$ pip install django-ldap
~~~

2. Add "django_ldap" to your INSTALLED_APPS setting  
![installed_apps](https://raw.githubusercontent.com/strife-dev/django-ldap/main/documentation/img/installed_apps.png)

3. Add "django_ldap.backend.LDAPBackend" as an AUTHENTICATION_BACKENDS i.e.

```python
AUTHENTICATION_BACKENDS = [
    # ...
    'django_ldap.backend.LDAPBackend',
    # ...
]
```
![auth_backend](https://raw.githubusercontent.com/strife-dev/django-ldap/main/documentation/img/auth_backend.png)

4. Add the following configuration items to your settings.py as a minimum
```LDAP_HOST = "LDAP://<LDAP_SERVER_HOSTNAME>:<LDAP_SERVER_PORT>"```
## Additional Configuration Parameters

| Key Name              | Value Default          | Description                                                                                                                                              | Sample                                             |
|:----------------------|:-----------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------|
| LDAP_HOST             | "ldap://localhost:389" | LDAP Server FDQN                                                                                                                                         | `LDAP_HOST = "LDAP://localhost:389"`               |
| LDAP_TLS              | False                  | Whether to enable secure ldap (note, LDAP and appropriate port should be used in hostname i.e. 636)                                                      | `LDAP_TLS = False`                                 |
| LDAP_TLS_VALIDATE     | ssl.CERT_OPTIONAL      | If TLS is enabled then the relevant SSL verification code required (either via ssl enum or int)                                                          | `LDAP_TLS_VALIDATE = ssl.CERT_OPTIONAL`            |
| LDAP_TLS_VERSION      | ssl.PROTOCOL_TLSv1     | If TLS is enabled then the relevant TLS version must be provided                                                                                         | `LDAP_TLS_VERSION = False`                         |
| LDAP_BIND_USER_DN     | None                   | DN of Auth User (None for anonymous)                                                                                                                     | `LDAP_BIND_USER_DN = "cn=admin,dc=example,dc=com"` |
| LDAP_BIND_PASS        | None                   | Auth User Password (ignored if user DN is None)                                                                                                          | `LDAP_BIND_PASS = "Foo"`                           |
| LDAP_PAGE_SIZE        | 500                    | Number of elements to return per paged result                                                                                                            | `LDAP_PAGE_SIZE = 500`                             |
| LDAP_SEARCH_ROOT      | "dc=example,dc=com"    | LDAP tree root (Used for search when left as default)                                                                                                    | `LDAP_SEARCH_ROOT = "dc=example,dc=com"`           |
| LDAP_ENABLE_REF_CACHE | True                   | Enables Referral Cache                                                                                                                                   | `LDAP_ENABLE_REF_CACHE = True`                     |
| LDAP_SET_LDAP3_ARGS   | {}                     | Allows LDAP3 global parameters to be updated (see [LDAP3 Global Config](https://ldap3.readthedocs.io/en/latest/installation.html#global-configuration) ) | `LDAP_SET_LDAP3_ARGS = {'SOCKET_SIZE': 4096}`      |
| LDAP_AUTH_MODE        | None                   | NTLM for Active Directory, SASL for external cert or kerberos, leave as None for other cases, ldap3 will auto choose SIMPLE or ANONYMOUS auth            | `LDAP_AUTH_MODE = NTLM`                            |

## How to contribute
TBD

ensure black formatter is used

CHANGES
0.0.2 - 2021-06-07
------------------

* Support LDAP3 Dynamic Param Loading

0.0.1 - 2021-06-06
------------------

* Initial Commit - Structure

django_ldap/ldap.py
from contextlib import contextmanager

from ldap3 import Server, Tls, Connection
from ldap3.utils import config

from django_ldap.settings import logger, ldap_settings


@contextmanager
def ldap_connection(*args, **kwargs):
    logger.info("Creating LDAP connection")

    for param, value in ldap_settings.LDAP_SET_LDAP3_ARGS.items():
        config.set_config_parameter(param, value)

    tls_configuration = (
        Tls(
            validate=ldap_settings.LDAP_TLS_VALIDATE,
            version=ldap_settings.LDAP_TLS_VERSION,
        )
        if ldap_settings.LDAP_TLS
        else None
    )

    server = Server(
        ldap_settings.LDAP_HOST,
        use_ssl=ldap_settings.LDAP_TLS,
        tls=tls_configuration,
    )

    with Connection(
        server,
        user=ldap_settings.LDAP_BIND_USER_DN,
        password=ldap_settings.LDAP_BIND_PASS,
        auto_bind=True,
        authentication=ldap_settings.LDAP_AUTH_MODE,
    ) as conn:
        logger.info(f"Connection: {conn}")

        yield conn


django_ldap/settings.py
"""
Settings for LDAP framework are all namespaced in the DJANGO_LDAP setting.
For example, in your settings.py:
DJANGO_LDAP = {
    'LDAP_HOST': "ldap://localhost:389",
    'LDAP_TLS': True
    'LDAP_SET_LDAP3_ARGS': {
        'SOCKET_SIZE': 4096,
        'RESTARTABLE_TRIES': 30
    },
}
This module exposes an `ldap_setting` object used across this framework, user options set in the DJANGO_LDAP settings
option are taken preference over the defaults set in this file.

Implementation derived from Django Rest Framework code, https://github.com/encode/django-rest-framework/blob/master/rest_framework/settings.py
"""
import logging
import ssl

from django.conf import settings
from django.test.signals import setting_changed
from django.utils.module_loading import import_string

logger = logging.getLogger("django_ldap")

DEFAULTS = {
    "LDAP_DRIVER": "ldap3",  # currently only supports ldap3, eventually python-ldap
    "LDAP_HOST": "ldap://localhost:389",
    "LDAP_TLS": False,
    "LDAP_TLS_VALIDATE": ssl.CERT_OPTIONAL,
    "LDAP_TLS_VERSION": ssl.PROTOCOL_TLSv1,
    "LDAP_BIND_USER_DN": None,
    "LDAP_BIND_PASS": None,
    "LDAP_PAGE_SIZE": "500",
    "LDAP_SEARCH_ROOT": "dc=example,dc=com",
    "LDAP_ENABLE_REF_CACHE": True,
    "LDAP_SET_LDAP3_ARGS": {},
    "LDAP_AUTH_MODE": None,
}

# List of settings that may be in string import notation.
IMPORT_STRINGS = []

# List of settings that have been removed
REMOVED_SETTINGS = []


def perform_import(val, setting_name):
    """
    If the given setting is a string import notation,
    then perform the necessary import or imports.
    """
    if val is None:
        return None
    elif isinstance(val, str):
        return import_from_string(val, setting_name)
    elif isinstance(val, (list, tuple)):
        return [import_from_string(item, setting_name) for item in val]
    return val


def import_from_string(val, setting_name):
    """
    Attempt to import a class from a string representation.
    """
    try:
        return import_string(val)
    except ImportError as e:
        msg = "Could not import '%s' for LDAP setting '%s'. %s: %s." % (
            val,
            setting_name,
            e.__class__.__name__,
            e,
        )
        raise ImportError(msg)


class LDAPSettings:
    """
    A settings object that allows REST Framework settings to be accessed as
    properties. For example:
        from django_ldap.settings import ldap_settings
        print(ldap_settings.LDAP_HOST)
    Any setting with string import paths will be automatically resolved
    and return the class, rather than the string literal.
    Note:
    This is an internal class that is only compatible with settings namespaced
    under the DJANGO_LDAP name.
    """

    def __init__(self, user_settings=None, defaults=None, import_strings=None):
        if user_settings:
            self._user_settings = self.__check_user_settings(user_settings)
        self.defaults = defaults or DEFAULTS
        self.import_strings = import_strings or IMPORT_STRINGS
        self._cached_attrs = set()

    @property
    def user_settings(self):
        if not hasattr(self, "_user_settings"):
            self._user_settings = getattr(settings, "DJANGO_LDAP", {})
        return self._user_settings

    def __getattr__(self, attr):
        if attr not in self.defaults:
            raise AttributeError("Invalid setting: '%s'" % attr)

        try:
            # Check if present in user settings
            val = self.user_settings[attr]
        except KeyError:
            # Fall back to defaults
            val = self.defaults[attr]

        # Coerce import strings into classes
        if attr in self.import_strings:
            val = perform_import(val, attr)

        # Cache the result
        self._cached_attrs.add(attr)
        setattr(self, attr, val)
        return val

    def __check_user_settings(self, user_settings):
        SETTINGS_DOC = "https://github.com/strife-dev/django-ldap/blob/master/documentation/settings.md"
        for setting in REMOVED_SETTINGS:
            if setting in user_settings:
                raise RuntimeError(
                    "The '%s' setting has been removed. Please refer to '%s' for available settings."
                    % (setting, SETTINGS_DOC)
                )
        return user_settings

    def reload(self):
        for attr in self._cached_attrs:
            delattr(self, attr)
        self._cached_attrs.clear()
        if hasattr(self, "_user_settings"):
            delattr(self, "_user_settings")


ldap_settings = LDAPSettings(None, DEFAULTS, IMPORT_STRINGS)


def reload_settings(*args, **kwargs):
    setting = kwargs["setting"]
    if setting == "DJANGO_LDAP":
        ldap_settings.reload()


setting_changed.connect(reload_settings)
